enhancements/ingress/lb-subnet-selection-aws.md

```diff
@@ -0,0 +1,591 @@
+annotation updates to NLBs once it has been created. However, requiring a service
+deletion and recreation will effectuate these changes.
+
+### Install Config Proposal (Day 1)
+
+TODO: Install time configuration proposal
+
+### Workflow Description
+
+#### Day 1: Setting IngressController Subnets during Installation Workflow
+
+Setting the subnets for the default IngressController during installation:
+
+TODO: Workflow for install time configuration
+
+#### Day 2: Creating an IngressController with Subnets after Installation Workflow
+
+Creating an IngressController while specifying the subnets for the LoadBalancer-type
+service after installation:
+
+1. A cluster admin creates an IngressController with
+  `spec.endpointPublishingStrategy.loadBalancer.providerParameters.aws.subnets` specified.
+2. The ingress operator creates the LoadBalancer-type Service and sets the
+   `service.beta.kubernetes.io/aws-load-balancer-subnets` annotation with the subnets
+   provided by the user.
+
+#### Updating an existing IngressController with new Subnets Workflow
+
+Updating an existing IngressController with new subnets requires the service
+to be deleted and recreated:
+
+1. A cluster admin edits an IngressController's
+   `spec.endpointPublishingStrategy.loadBalancer.providerParameters.aws.subnets` field.
+2. The Ingress Operator sets the `LoadBalancerProgressing` condition to `Status: True`
+   on the IngressController and provides a message indicating to the cluster admin
+   that they must delete the service.
+3. The cluster admin deletes the LoadBalancer-type service indicated in the
+   `LoadBalancerProgressing` condition message.
+4. The Ingress Operator recreates the LoadBalancer-type service and configures the
+   `service.beta.kubernetes.io/aws-load-balancer-subnets` on the LoadBalancer-type
+   service, thus creating the load balancer with the desired subnets specified by the
+   cluster admin.
+
+#### Unmanaged Subnet Annotation Migration Workflow
+
+Migrating an unmanaged `service.beta.kubernetes.io/aws-load-balancer-subnets`
+service annotation to `spec.endpointPublishingStrategy.loadBalancer.providerParameters.aws.subnets`
+after upgrading to 4.y doesn't require a cluster admin to delete the LoadBalancer-type
+service:
+
+1. Cluster admin confirms that initially `service.beta.kubernetes.io/aws-load-balancer-subnets`
+   is set on the LoadBalancer-type service managed by the IngressController.
+2. The cluster admin upgrades OpenShift to v4.y and the service annotation is not
+   changed or removed.
+3. After upgrading, the IngressController will emit a `LoadBalancerProgressing` condition
+   with `Status: True` because the spec's `Subnets` does not equal the current annotation.
+4. Instead of recreating the LoadBalancer-type service, the cluster admin will set
+   `spec.endpointPublishingStrategy.loadBalancer.providerParameters.aws.subnets` to the
+   current value of the `service.beta.kubernetes.io/aws-load-balancer-subnets` service.
```

Comment:
I meant the case described above in "Updating an existing IngressController with new Subnets Workflow" chapter where the cluster admin is asked to delete the exposing service:

> 2. The Ingress Operator sets the `LoadBalancerProgressing` condition to `Status: True`
   on the IngressController and provides a message indicating to the cluster admin
   that they must delete the service.
   
Will the cluster admin be asked (via IngressController status) to add the subnets from the annotation to the spec field?

Reply:
Yes there will still be a `LoadBalancerProgressing` status message, however, it won't exactly tell them to do this workflow, this is what it will say:

```
  - message: "The IngressController subnets were changed from [\"\"] to [\"subnetA\", \"subnetB\", \"subnetC\"]. To effectuate this change, you must delete the service: `oc -n openshift-ingress delete svc/custom-default`; the service load-balancer will then be deprovisioned and a new one created. This will most likely cause the new load-balancer to have a different host name and IP address from the old one's. Alternatively, you can revert the change to the IngressController: `oc -n openshift-ingress-operator patch ingresscontrollers/default --type=merge --patch='{\"spec\":{\"endpointPublishingStrategy\":{\"type\":\"LoadBalancerService\",\"loadBalancer\":{\"providerParameters\":{\"type\":\AWS\",\"aws\":{\"type\":\"Classic\",\"subnets\":[]}}}}}}'"
    reason: OperandsProgressing
    status: "True"
    type: LoadBalancerProgressing
```

The upgrade case (spec == `[]` and annotation == `<foo>`) is the same scenario as when a cluster admin clears/removes the subnets from the spec after setting the spec once. In that case, we do want them to delete and recreate the service. So the upgrade case isn't uniquely detectable.

The last part of the message:
```
Alternatively, you can revert the change to the IngressController: `oc -n openshift-ingress-operator patch ingresscontrollers/default --type=merge --patch='{\"spec\":{\"endpointPublishingStrategy\":{\"type\":\"LoadBalancerService\",\"loadBalancer\":{\"providerParameters\":{\"type\":\AWS\",\"aws\":{\"type\":\"Classic\",\"subnets\":[]}}}}}}'"
```
Will actually give them the exact command to set the spec back to the annotation. Though, yes, it might be super clear that that's what they should run.

I didn't add a specific message describe what to do if you are in a upgrade case, because technically we don't support unmanaged `service.beta.kubernetes.io/aws-load-balancer-subnets` on the service. I think it could be confusing for these cluster admins upgrading with an unmanaged/unsupported annotation, but I was thinking they would probably figure it out by reading the status message. Worse comes to worse, they follow the instructions and delete their service, it removes their subnet annotation, and they do some research and find out we added the `Subnets` field.

Does that make sense? Yes, they will get a status message, but it won't exactly tell them how to do this workflow, but I don't necessarily feel strongly compelled to handle that since this is an unsupported situation.

Reply:
> So the upgrade case isn't uniquely detectable.

Right.

> The upgrade case (spec == [] and annotation == <foo>) is the same scenario as when a cluster admin clears/removes the subnets from the spec after setting the spec once. In that case, we do want them to delete and recreate the service.

So the user can't re-enable the auto discovery?

About the message:
1. It's for the case when the annotation has `"subnetA", "subnetB", "subnetC"` and the spec has empty list? Or the opposite?

> The IngressController subnets were changed from [\"\"] to [\"subnetA\", \"subnetB\", \"subnetC\"]

2. At the end the status message says to delete the service to proceed with the change:

> To effectuate this change, you must delete the service

Reply:
> So the user can't re-enable the auto discovery?

They can indeed. They just have to delete and recreate the service any time they change the spec `Subnets`.

> It's for the case when the annotation has "subnetA", "subnetB", "subnetC" and the spec has empty list? Or the opposite?

In this message example, it's the case that the annotation was changed to `"subnetA", "subnetB", "subnetC"` and the spec has the empty list.

> At the end the status message says to delete the service to proceed with the change:

Correct. So going back to the update scenario, I agree this may be confusing for cluster admins that are using an unsupported annotation. But I don't feel strongly about changing the wording because it's an supported configuration, but I'm open to differing opinions. I think the main concern for handling unmanaged annotations was that we don't immediately break their clusters upon upgrade, which is satisfied.

Reply:
If it's unsupported it's probably better not to mention it all and tell them to manually remove the annotation before upgrading.